[Swift] Add initial support and tests for Swift parseable interface files (.swiftinterface)

nathawes opened the pull request:
This patch registers a `ParseableInterfaceModuleLoader` with LLDB's `SwiftASTContext` so it can load .swiftinterface files and adds tests that we can still debug executables that link to and use types from libraries with only a .swiftinterface (i.e. no .swiftmodule) and no debug info.

Fixes rdar://problem/43905961

Commit:
[Swift] Add initial support and tests for Swift parseable interface files (.swiftinterface)

This patch (1) registers a ParseableInterfaceModuleLoader with LLDB's
SwiftASTContext so it can load .swiftinterface files, and (2) adds tests that
we can still debug executables linking to libraries that only have a
.swiftinterface (i.e. no .swiftmodule).

Fixes rdar://problem/43905961

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

Commit:
[Swift] Fix the tests for .swiftinterface files to run on Linux

Add support to Makefile.rules for building Swift shared libraries on Linux.
Add support for not linking in the wrapped swiftmodule.
Update lit config with %target-shared-library-suffix.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci clean test linux

nathawes commented:
@swift-ci clean test os x

nathawes commented:
@swift-ci clean test os x platform

nathawes commented:
@swift-ci clean test macOS

nathawes commented:
@swift-ci clean test

nathawes commented:
@swift-ci clean test

nathawes commented:
@swift-ci test

nathawes commented:
@swift-ci test

nathawes commented:
@swift-ci test

nathawes commented:
@swift-ci test

nathawes commented:
@swift-ci test OS X

nathawes commented:
@swift-ci test

adrian-prantl commented on lit/SwiftREPL/SwiftInterface.test:
```diff
@@ -0,0 +1,17 @@
+// RUN: rm -rf %t
+// RUN: mkdir %t
+// RUN: %target-swiftc -module-name A -emit-parseable-module-interface-path %t/A.swiftinterface -emit-library -o %t/libA%target-shared-library-suffix %S/Inputs/A.swift
+// RUN: %lldb --repl="-I%t -L%t -lA -enable-parseable-module-interface" < %s | FileCheck %s
```
It might be good to make sure that it is impossible to accidentally import the library from source code, perhaps by moving away the sources.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/swift_interface_shared/TestSwiftInterfaceNoDebugInfo.py:
```diff
@@ -0,0 +1,135 @@
+
+        # Set a breakpoint in and launch the main executable
+        (target, process, thread, bkpt) = lldbutil.run_to_source_breakpoint(self,
+                "break here", self.main_source_spec,
```
perhaps just construct the filespec inline here?

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/swift_interface_shared/TestSwiftInterfaceNoDebugInfo.py:
```diff
@@ -0,0 +1,135 @@
+                exe_name="main")
+
+        self.frame = thread.frames[0]
+        self.assertTrue(self.frame, "Frame 0 is valid.")
```
this can be removed run_to_source_breakpoint already checks this.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/swift_interface_shared/TestSwiftInterfaceNoDebugInfo.py:
```diff
@@ -0,0 +1,135 @@
+        lldbutil.check_expression(self, self.frame, "y.magnitudeSquared", "404", use_summary=False)
+        lldbutil.check_expression(self, self.frame, "MyPoint(x: 1, y: 2).magnitudeSquared", "5", use_summary=False)
+
+        # Check the module cache was populated with the .swiftmodule files of
```
Can you spell this "Swift module cache" to avoid confusing with the clang module cache?

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/swift_interface_shared/TestSwiftInterfaceNoDebugInfo.py:
```diff
@@ -0,0 +1,135 @@
+
+def is_old(file):
+    """Checks the modified time of the given file matches the timestamp set my make_old"""
+    return os.stat(file).st_mtime == OLD_TIMESTAMP
```
This may be a portability issue, but we'll see.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/swift_interface_shared/libs/Makefile:
```diff
@@ -0,0 +1,17 @@
+# this test
+SWIFTFLAGS=-Onone -Xfrontend -serialize-debugging-options -enable-parseable-module-interface
+
+# don't include the wrapped .swiftmodule on linux to make sure we actually use the .swiftinterface
```
Add comment: `the issue here is that because the .swiftmodule is being wrapped and linked into the dylib, LLDB will find it there and load it before the .swiftinterface loading path can be exercised.`

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/swift_interface_static/TestSwiftInterfaceStaticNoDebugInfo.py:
```diff
@@ -0,0 +1,136 @@
+
+        # Set a breakpoint in and launch the main executable
+        (target, process, thread, bkpt) = lldbutil.run_to_source_breakpoint(self,
+                "break here", self.main_source_spec,
```
SBFile()

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/swift_interface_static/TestSwiftInterfaceStaticNoDebugInfo.py:
```diff
@@ -0,0 +1,136 @@
+                exe_name="main")
+
+        self.frame = thread.frames[0]
+        self.assertTrue(self.frame, "Frame 0 is valid.")
```
redundant

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/swift_interface_static/TestSwiftInterfaceStaticNoDebugInfo.py:
```diff
@@ -0,0 +1,136 @@
+        self.runCmd("expr import CC")
+
+        # Not clear why this doesn't work here, when it does with a dylib...
+        lldbutil.check_expression(self, self.frame, "Baz.baz()", None, use_summary=False)
```
Could you please file a bug report for this?

jrose-apple replied:
Static libraries just can't be loaded dynamically; it's not related to this new feature.

adrian-prantl commented on source/Symbol/SwiftASTContext.cpp:
```diff
@@ -2950,6 +2950,21 @@ swift::ASTContext *SwiftASTContext::GetASTContext() {
       m_ast_context_ap->addModuleLoader(std::move(serialized_module_loader_ap));
     }
 
+    // Install the parseable interface module loader
```
`.` at the end please

adrian-prantl reviewed (commented):
Looks almost good! Thanks a lot.

Commit:
[Swift] Minor updates to the tests for .swiftinterface support based on review feedback

nathawes commented:
@swift-ci test

nathawes commented:
@swift-ci test OS X

nathawes commented:
@swift-ci test OS X

jrose-apple commented on source/Symbol/SwiftASTContext.cpp:
```diff
@@ -2950,6 +2950,21 @@ swift::ASTContext *SwiftASTContext::GetASTContext() {
       m_ast_context_ap->addModuleLoader(std::move(serialized_module_loader_ap));
     }
 
+    // Install the parseable interface module loader.
```
I think this one has to go before the serialized one in order to get the correct fallback behavior when the binary swiftmodule is invalid / unreadable / out of date. I guess there should be a test for that too!

nathawes replied:
Good catch – thank you! Moved it up and added a version of the tests that adds empty .swiftmodule files corresponding to each of the libs beforehand.

Commit:
[Swift] Install the ParseableInterfaceModuleLoader before the SerializedModuleLoader for correct fallback behavior

Also added a test that writes out empty .swiftmodule files and checks we
fallback to using the .swiftinterface file when we fail to load them.

nathawes commented:
@swift-ci test

nathawes commented:
Thanks @jrose-apple and @adrian-prantl! I think I've addressed the feedback now. Would you mind taking another look?

jrose-apple reviewed (approved):
These look good. I suggest another test where the libraries *do* have debug info (and are processed with dsymutil). In that case we should be able to get private stuff out of AA and BB (because they were loaded into the process) but not CC (because the user imported that themselves). Or maybe LLDB will pick up the debug info for CC too; that's okay as well.

nathawes commented:
Going to land this and add a version of the tests with debug info in a follow-up PR with another test I'm working on.
